cf/latest-snapshot/src/index.ts

```diff
@@ -7,31 +9,63 @@ function basename(path: String) {
 }
 
 // Directly fetch the data for the latest snapshot of a given chain (eg. calibnet or mainnet)
-async function get_latest(env: Env, chain: string): Promise<Response> {
+async function get_latest(req_headers: Headers, env: Env, chain: string): Promise<Response> {
 	const listed = await env.FOREST_ARCHIVE.list({ prefix: chain + "/latest/" });
 	let latest = listed.objects.at(-1);
 	if (latest == null) {
 		return new Response(`No latest snapshot found ${chain}`, {
 			status: 404,
 		});
 	} else {
-		// Should we support range queries?
-		const object = await env.FOREST_ARCHIVE.get(latest.key);
+		const object = await env.FOREST_ARCHIVE.get(latest.key, {
+			range: req_headers,
+			onlyIf: req_headers,
+		});
+
 		if (object === null) {
 			return new Response('No latest snapshot found', {
 				status: 404,
 			});
 		}
+
 		const headers = new Headers();
+
+		// If the client requested a range, then we need to set the Content-Range header.
+		// The `parseRange` returns an error code if the range is not satisfiable (or not specified),
+		// so we can handle the range request only if it returns an array.
+		let status = 200;
+		let range = parseRange(object.size, req_headers.get("range") || "");
+		if (Array.isArray(range)) {
+			// R2Object doesn't support multiple ranges, so we only use the first one.
+			// Throw an error if there are more than one range.
+			if (range.length > 1) {
+				return new Response('Multiple ranges are not supported', {
+					status: 416, // Range Not Satisfiable
+				});
+			}
+
+			let r = range[0];
+			headers.set("Content-Range", `${r.type} ${r.start}-${r.end}/${object.size}`);
+			status = 206; // Partial Content
+		}
+
 		object.writeHttpMetadata(headers);
 		headers.set('etag', object.httpEtag);
-		let encoded_name = encodeURIComponent(basename(object.key));
+		const encoded_name = encodeURIComponent(basename(object.key));
 		// Tell browsers and aria2c which filename to use. For 'wget', you have to use `--trust-server-names`.
-		headers.set('content-disposition', `attachment; filename*=UTF-8''${encoded_name}; filename="${encoded_name}"`);
+		headers.set('Content-Disposition', `attachment; filename*=UTF-8''${encoded_name}; filename="${encoded_name}"`);
 
-		return new Response(object.body, {
-			headers,
-		});
+		if ('body' in object) {
+			return new Response(object.body, {
```

Comment:
I feel there should be an easier way of doing this but I know no JS.

Reply:
I found no easier way (though many more difficult and esoteric ways which strangely resembling C++ template metaprogramming)